R/verb-copy-to.R

```diff
@@ -87,6 +87,64 @@ copy_to.src_sql <- function(dest, df, name = deparse(substitute(df)),
   invisible(out)
 }
 
+#' @export
+db_values <- function(con, df, name = deparse(substitute(df))) {
+  if (ncol(df) == 0) {
+    abort("`df` needs at least one column.")
+  }
+
+  if (!inherits(df, "data.frame")) {
+    abort("`df` needs to be a data.frame.")
+  }
+
+  tbl(
+    src = con,
+    from = sql_values(con, df),
+    vars = colnames(df)
+  )
+}
+
+sql_values <- function(con, df) {
+  UseMethod("sql_values")
+}
+
+#' @export
+sql_values.DBIConnection <- function(con, df) {
+  sql_values_clause(con, df)
+}
+
+sql_values_clause <- function(con, df, row = FALSE) {
+  sim_data <- rep_named(colnames(df), list(NULL))
+  cols_clause <- escape(sim_data, con = con, parens = FALSE)
+  null_row_clause <- build_sql("SELECT ", cols_clause, " WHERE false", con = con)
+
+  if (nrow(df) == 0) {
+    return(null_row_clause)
+  }
+
+  values <- purrr::map(df, escape, con = con, collapse = NULL, parens = FALSE)
+  rows <- purrr::pmap(
+    unname(values),
+    function(...) {
+      dots <- list(...)
+      out <- sql_vector(dots, con = con, collapse = ", ", parens = TRUE)
+      if (row) {
+        out <- sql(paste0("ROW", out))
```

Comment:
Is this missing a space?

```suggestion
        out <- sql(paste0("ROW ", out))
```

Reply:
No, it is intended this way: https://dev.mysql.com/doc/refman/8.0/en/values.html

```sql
VALUES ROW(1,-2,3), ROW(5,7,9), ROW(4,6,8);
```